feat: remove deprecated direct execution from k6 cloud command

moko-poi opened the pull request:
## What?

Remove the deprecated `k6 cloud script.js` direct execution syntax and enforce the use of subcommands (`run`, `login`, `upload`) for all cloud operations.

**Key changes:**
- Updated `exactCloudArgs()` to reject direct script execution and require subcommands
- Modified error messages to guide users to the correct `k6 cloud run` syntax
- Updated help text and examples to reflect v2.0.0 API changes
- Cleaned up deprecated references in command descriptions
- Renamed internal variables from `deprecatedCloudCmd` to `cloudCmd` for clarity
- Updated all tests to use the new subcommand syntax

## Why?

The `k6 cloud script.js` command has been deprecated for several versions in favor of `k6 cloud run script.js`. This change:

1. **Simplifies the API** by enforcing a consistent subcommand structure
2. **Reduces confusion** between different cloud operations (run, login, upload)
3. **Aligns with planned breaking changes** for v2.0.0 as outlined in issue #5116
4. **Improves user experience** with clearer, more explicit command syntax

Users attempting to use the old syntax now receive clear guidance on how to migrate to the new commands.

## Related Issue

https://github.com/grafana/k6/issues/5116

## Migration Guide

**Before (deprecated):**
```bash
k6 cloud script.js
k6 cloud archive.tar
```

**After (v2.0.0):**
```bash
k6 cloud run script.js
k6 cloud run archive.tar
```

All other cloud functionality remains unchanged:
```bash
k6 cloud login
k6 cloud upload script.js
```

moko-poi committed:
feat: remove direct execution from k6 cloud command

moko-poi committed:
test: migrate cloud tests to use 'run' subcommand

moko-poi commented:
Hi @mstoykov @AgnesToulet , this PR has been ready for review for 2 weeks now. Could someone please take a look when you have a chance? This is part of the v2.0.0 breaking changes milestone, so I'd appreciate any feedback or approval to move this forward.

cc @codebien (issue author)

codebien reviewed (changes requested):
https://github.com/grafana/k6/pull/5134#pullrequestreview-3257269850

moko-poi commented:
https://github.com/grafana/k6/pull/5134#issuecomment-3328229373

mstoykov commented:
Hi @moko-poi do you want to reopen this PR? It is now possible to merge it as we are now wokring directly on v2.

moko-poi committed:
fix: update TestBinaryNameHelpStdout to expect 'cloud run' subcommand syntax

The test was checking for the deprecated 'k6 cloud script.js' in help
output, but since direct execution has been removed, update to expect
'k6 cloud run script.js' instead.

moko-poi commented:
@mstoykov Thanks for the heads up! I've reopened the PR and rebased it onto the latest master.

Changes made
- Rebased onto current master
- Resolved merge conflicts in `internal/cmd/cloud.go`
- Fixed `TestBinaryNameHelpStdout` to expect `cloud run script.js` instead of the deprecated `cloud script.js` syntax (as flagged by @codebien)

Ready for review!

codebien commented on internal/cmd/cloud_run.go:
```diff
@@ -63,8 +63,8 @@ func getCmdCloudRun(cloudCmd *cmdCloud) *cobra.Command {
 	cloudRunCmd := &cmdCloudRun{
-		deprecatedCloudCmd: cloudCmd,
-		runCmd:             runCmd,
+		cloudCmd: cloudCmd,
```
```suggestion
		parentCloudCmd: cloudCmd,
```
or `rootCloudCmd`.

I guess this architecture now is not anymore required, it would be useful to refactor this command in a follow-up pull request. @moko-poi would you be interested into give it a try?

moko-poi replied:
Thanks for the suggestion! I've applied the rename to `parentCloudCmd` in f39b93f.

And yes, I'd be happy to take on the follow-up refactoring PR to clean up this architecture. I'll create a separate PR for that.

mstoykov commented:
One integration test seems to be failing @moko-poi

moko-poi committed:
fix: update TestRootCommandHelpDisplayCommands to match new cloud command description

moko-poi committed:
refactor: rename cloudCmd to parentCloudCmd in cmdCloudRun

Apply suggested change from code review to clarify the relationship
between the cloud run subcommand and its parent cloud command.

codebien commented on internal/cmd/cloud.go:
```diff
@@ -367,31 +369,36 @@ func getCmdCloud(gs *state.GlobalState) *cobra.Command {
+- "upload": Upload a test script to Grafana Cloud k6 without running it
+
+The direct usage of "k6 cloud script.js" has been removed in v2.0.0.
+Please use "k6 cloud run script.js" instead.`,
```
I don't think we need this. I guess Claude got very verbose here.

moko-poi replied:
Fixed! Removed the verbose subcommand listing and migration notice.

moko-poi committed:
fix: simplify cloud command Long description

Remove verbose subcommand listing and migration notice from the
cloud command's Long description. Subcommands are already visible
via --help output.

codebien commented on internal/cmd/cloud.go:
```diff
@@ -433,6 +432,28 @@ Use "{{.CommandPath}} [command] --help" for more information about a command.
 	return cloudCmd
 }
 
+func exactCloudArgs() cobra.PositionalArgs {
+	return func(_ *cobra.Command, args []string) error {
+		const baseErrMsg = `the "k6 cloud" command requires a subcommand such as "run", "login", or "upload"`
+
+		if len(args) == 0 {
+			return fmt.Errorf(baseErrMsg + "; " + "received no arguments")
+		}
+
+		hasSubcommand := args[0] == "run" || args[0] == "login" || args[0] == "upload"
+		if !hasSubcommand {
+			return fmt.Errorf(
+				baseErrMsg+"; "+
+					`direct script execution has been removed in v2.0.0. `+
+					`To run "%s", use: k6 cloud run %s`,
+				args[0], args[0],
+			)
+		}
+
+		return nil
+	}
+}
```
Are you sure this is needed? Can you give me a command that triggers this, please?

codebien commented on internal/cmd/cloud.go:
```diff
@@ -367,31 +369,28 @@ func getCmdCloud(gs *state.GlobalState) *cobra.Command {
 
-  # Run a test archive in Grafana Cloud
-  $ {{.}} cloud run archive.tar`[1:])
+  # Run a k6 archive in Grafana Cloud k6
+  $ {{.}} cloud run archive.tar
+
+  # Upload a test script to Grafana Cloud k6
+  $ {{.}} cloud upload script.js`[1:])
 
 	cloudCmd := &cobra.Command{
-		Use:     "cloud",
-		Short:   "Run and manage Grafana Cloud tests",
-		Long:    "Run and manage tests in Grafana Cloud.",
+		Use:   "cloud",
+		Short: "Run a test on the cloud",
+		Long:  `Manage Grafana Cloud k6 tests.`,
+		Args:    exactCloudArgs(),
```
I don't think we expect to change any text here. If there is a reason I'm happy to add it, but if this is not adding value then I'm for rollback them and reduce the number of lines changed.

codebien reviewed (changes requested):
A few more request changes.

Generally speaking, please ensure that you only include changes relevant to the current pull request. If there are any low-hanging fruit opportunities, they are welcome. But try not to just grab every AI suggestion without first checking if it is actually needed and relevant.

moko-poi commented:
Sorry about that. I'll take more time to review each change carefully and make sure everything is scoped to the PR going forward.

codebien commented:
Hey @moko-poi :wave:
I wanted to follow up here, since the v2 release date is approaching, are still interested into contributing to get this included?

codebien commented:
Hi :wave:
As previously announced, this change must be included. We have opened a new pull request to address this #5912. We apologize for the inconvenience and appreciate your contribution here.
